components/pkg-export-helm/src/chart.rs

```diff
@@ -0,0 +1,202 @@
+        let chartfile = ChartFile::new(&name, version, description)?;
+
+        Chart::new_for_manifest(manifest, name, chartfile, ui)
+    }
+
+    fn new_for_manifest(
+        manifest: Manifest,
+        name: String,
+        chartfile: ChartFile,
+        ui: &'a mut UI,
+    ) -> Result<Self> {
+        let main = json!({
+            "metadata_name": "{{.Values.metadataName}}",
+            "habitat_name": "{{.Values.habitatName}}",
+            "image": "{{.Values.imageName}}",
+            "count": "{{.Values.instanceCount}}",
+            "service_topology": "{{.Values.serviceTopology}}",
+            "service_group": manifest.service_group.clone().map(|_| "{{.Values.serviceGroup}}"),
+            "config_secret_name": manifest.config_secret_name
+                .clone()
+                .map(|_| "{{.Values.configSecretName}}"),
+                "ring_secret_name": manifest.ring_secret_name
+                .clone()
+                .map(|_| "{{.Values.ringSecretName}}"),
+                "bind": manifest.binds,
+        });
+
+        let mut values = Values::new();
+        values.add_entry("metadataName", &manifest.metadata_name);
+        values.add_entry("habitatName", &manifest.habitat_name);
+        values.add_entry("imageName", &manifest.image);
+        values.add_entry("instanceCount", &manifest.count.to_string());
+        values.add_entry("serviceTopology", &manifest.service_topology.to_string());
+        if let Some(ref group) = manifest.service_group {
+            values.add_entry("serviceGroup", group);
+        }
+        if let Some(ref name) = manifest.config_secret_name {
+            values.add_entry("configSecretName", name);
+        }
+        if let Some(ref name) = manifest.ring_secret_name {
+            values.add_entry("ringSecretName", name);
+        }
+
+        let mut binds = Vec::new();
+        let mut i = 0;
+        for bind in &manifest.binds {
+            let split: Vec<&str> = bind.split(":").collect();
+            if split.len() < 3 {
+                return Err(Error::InvalidBindSpec(bind.to_string()).into());
+            }
+
+            let name_var = format!("bindName{}", i);
+            let service_var = format!("bindService{}", i);
+            let group_var = format!("bindGroup{}", i);
+            i = i + 1;
+
+            values.add_entry(&name_var, split[0]);
+            values.add_entry(&service_var, split[1]);
+            values.add_entry(&group_var, split[2]);
+
+            let json = json!({
+                "name": format!("{{{{.Values.{}}}}}", name_var),
+                "service": format!("{{{{.Values.{}}}}}", service_var),
+                "group": format!("{{{{.Values.{}}}}}", group_var),
+            });
+
+            binds.push(json);
+        }
+
+        let manifest_template = ManifestJson {
+            main: main,
+            binds: binds,
+        };
+        let habitat_name = manifest.habitat_name;
+
+        Ok(Chart {
+            name,
```

Comment:
Is this some magic where compiler compares the variable name to the name of the field in `Chart` and does the right thing? Or were you just careful to set it in correct order? If the latter, then I'd prefer to have `name: name,` then.

Reply:
yes that's a Rust feature: If the names are the same, you can use the simpler format. I double very much the order is important since otherwise, the names being the same rule doesn't apply.